homeassistant/components/keba/__init__.py

```diff
@@ -0,0 +1,160 @@
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'keba'
+SUPPORTED_COMPONENTS = ['binary_sensor', 'sensor', "lock"]
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional('rfid', default='00845500'): cv.string,
+        vol.Optional('failsafe', default=False): cv.boolean,
+        vol.Optional('failsafe_timeout', default=30): cv.positive_int,
+        vol.Optional('failsafe_fallback', default=6): cv.positive_int,
+        vol.Optional('failsafe_persist', default=0): cv.positive_int,
+        vol.Optional('refresh_interval', default=5): cv.positive_int,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+_SERVICE_MAP = {
+    'request_data': 'request_data',
+    'set_energy': 'async_set_energy',
+    'set_current': 'async_set_current',
+    'authorize': 'async_start',
+    'deauthorize': 'async_stop',
+    'enable': 'async_ena1',
+    'disable': 'async_ena0',
+    'set_failsafe': 'async_set_failsafe'
+}
+
+
+async def async_setup(hass, config):
+    """Check connectivity and version of KEBA charging station."""
+    host = config[DOMAIN][CONF_HOST]
+    rfid = config[DOMAIN]['rfid']
+    refresh_interval = config[DOMAIN]['refresh_interval']
+
+    keba = KebaHandler(hass, host, rfid, refresh_interval)
+    hass.data[DOMAIN] = keba
+
+    # Register services to hass
+    async def async_execute_service(call):
+        """Execute a service to KEBA charging station.
+
+        This must be a member function as we need access to the hass
+        object here.
+        """
+        function_name = _SERVICE_MAP[call.service]
+        function_call = getattr(keba, function_name)
+        hass.async_create_task(function_call(call.data))
+
+    # activate failsafe mode if enabled
+    failsafe = config[DOMAIN]['failsafe']
+    timeout = config[DOMAIN]['failsafe_timeout'] if failsafe else 0
+    fallback = config[DOMAIN]['failsafe_fallback'] if failsafe else 0
+    persist = config[DOMAIN]['failsafe_persist'] if failsafe else 0
+    try:
+        hass.loop.create_task(keba.set_failsafe(timeout, fallback, persist))
+    except ValueError as e:
+        _LOGGER.warning("Could not set failsafe mode " + str(e))
+
+    # Register services
+    for service in _SERVICE_MAP:
+        hass.services.async_register(DOMAIN, service, async_execute_service)
+
+    for domain in SUPPORTED_COMPONENTS:
+        hass.async_create_task(
+            discovery.async_load_platform(hass, domain, DOMAIN, {}, config))
+
+    return True
+
+
+class KebaHandler(KebaKeContact):
+    """Representation of a KEBA charging station connection."""
+
+    def __init__(self, hass, host, rfid, refresh_interval):
+        """Constructor."""
+        super().__init__(host)
+
+        self._update_listeners = []
+        self._hass = hass
+        self.rfid = rfid
+
+        # Ensure at least 5 seconds delay
+        self._refresh_interval = max(5, refresh_interval)
```

Comment:
According to the documentation "The minimum waiting time between the scheduled repetitions of any UPDcommand is defined as follows: t_COM_pause = 5 s". - https://www.keba.com/download/x/fff263235e/kecontactp30udp_pgen.pdf

In addition, I reverse-engineered the offical android app and it is using 5 seconds updates as well. Thus I ensure this limit with this line of code.

Reply:
Ok, it was just a question whether it makes sense to poll it so often, I'm unsure what's the current stance on this regard (as usually `update()` and `async_update()` are scheduled automatically by homeassistant to orchestrate this).

Reply:
The issue is that the data for the different sensors are bundled within mainly two requests to the charging station (states with "report 2" and for voltage and current "report 3"). Therefore I copied the concept of bimmer-connected where also all data is fetched from one instance periodically and the sensors get the data from that one because they are added as listener with the add_update_listener function.
The issue with the fetching interval is, that the connection is udp, so I don't get an answer if a command was successfully executed, thus I need to observe the state changes. In the android app a request interval of 5s is default and 1s after submitting a command. I think I will adopt this concept to fetch state changes in a smaller interval after sending a command. The regular interval can be set as you like with a minimum of 5s as from the developers manual.

Reply:
Edit: I integrated a fast polling interval after commands were executed (4 times 2 seconds). Afterwards the normal interval from the config is used again.

Reply:
Please extract the five seconds into a constant, so it's clear what the polling frequency is.

Reply:
done I also extracted the MAX_FAST_POLLING_COUNT to a variable (=4)